### ISSUE TITLE
Queued Builds

### ISSUE THREAD

**Issue by disturbing** [qeesung/mq-pipeline-notifier-plugin#1]
Kudos for open sourcing this!

Feature request - possible to get an event for queued pipelines and canceled/aborted pipelines?

**Comment by qeesung** [comment]
Of course. Just extend the `RunListener` and override the `onCompleted`, `onStarted`,`onDeleted` method, we can get an event for queued pipelines and canceled/aborted pipelines. I have on time work on this recently, welcome to pull request

**Comment by disturbing** [comment]
Ok, won't be able to work on this until the last week of October as easy as
it sounds. But we're implementing this into our core ci/cd so once it's
running next week, will be easier to test once we have this running in dev.

I'll keep you posted.

But just to be sure. On start and on delete for the runner is queue and
cancel job to be clear? I don't have exp on jenkins plugins :)

**Comment by qeesung** [comment]
In Jenkins, start event is not same to queued event, start event is triggered after job left the queue.

If we want to catch the event that a job is queued or dequeued,  extend the [QueueListener](http://javadoc.jenkins.io/hudson/model/queue/QueueListener.html) class, and override the methods:

```java
@Extension
public class QueueListenerImpl extends QueueListener{
    
    @Override
    public void onEnterWaiting(Queue.WaitingItem wi) { // queued

    }

    @Override
    public void onLeft(Queue.LeftItem li) { // dequeued

    }
}
```

To catch the start, deleted, completed(success, failed, aborted,...) event, just extend the [RunListener](http://javadoc.jenkins.io/hudson/model/listeners/RunListener.html):

```java
@Extension
public class RunListenerImpl extends RunListener<Run> {
    @Override
    public void onCompleted(Run run, @Nonnull TaskListener listener) { // job finished

    }

    @Override
    public void onStarted(Run run, TaskListener listener) { // job started

    }

    @Override
    public void onDeleted(Run run) { // job deleted
    
    }
}
```
